add special case for findlast on tuples of length >= 32, fixes #45117

pcjentsch opened the pull request:
This PR fixes issue #45117.

The problem was that `findfirst` has a special case `_findfirst_loop` for tuples of length >= 32, whereas `findlast` always call `findfirst(f, reverse(t))`. Since `reverse(t)` involves a splat, performance rapidly drops at 32 elements.

This PR simply adds a corresponding method `_findlast_loop` that loops over tuples of length >= 32 instead of splatting.

<details>
<summary> Benchmarks </summary>

```
using BenchmarkTools

symvector = collect(Symbol.("a" * string(i) for i = 1:36));
symtuple = Tuple(symvector);
firstsym = :a1;
lastsym = :a36;
@btime findlast(==($lastsym), $symtuple);
@btime findlast(==($lastsym), $(symtuple[1:32]));
@btime findlast(==($lastsym), $(symtuple[1:33]));

# before PR
# 11.940 μs (72 allocations: 6.47 KiB)
# 125.766 ns (0 allocations: 0 bytes)
# 10.490 μs (66 allocations: 5.53 KiB)

# after PR
# 2.349 ns (0 allocations: 0 bytes)
# 9.279 ns (0 allocations: 0 bytes)
# 9.699 ns (0 allocations: 0 bytes)
```

</details>

pcjentsch committed:
add special case for findlast on tuples of length >= 32, fixes #45117

KristofferC commented:
Just to check, are there already tests for tuples with both fewer and more elements than 32 for `findlast`?

pcjentsch committed:
fix bug in findlast, add tests for findfirst and findlast for tuples of length >= 32

pcjentsch commented:
> Just to check, are there already tests for tuples with both fewer and more elements than 32 for `findlast`?

There aren't any for `findlast` or `findfirst` for long tuples. I have added some, and also fixed a mistake in my code.
